File: apps/frontend/src/lib/api/permissions.ts
import { createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import { AdminPermissionService } from '$lib/pb/chatto/admin/v1/permissions_connect';
import {
  PermissionDecision,
  PermissionScopeKind,
  type PermissionMatrixCell as APIPermissionMatrixCell,
  type PermissionMatrixScope as APIPermissionMatrixScope,
  type RolePermissionMatrix as APIRolePermissionMatrix,
  type TierRole as APITierRole,
  type TierRoles as APITierRoles,
  type UserPermissionMatrix as APIUserPermissionMatrix
} from '$lib/pb/chatto/admin/v1/permissions_pb';

export type PermissionAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
};

export type PermissionState = 'allow' | 'deny' | 'neutral';
export type MatrixDecision = 'ALLOW' | 'DENY' | 'NONE';
export type MatrixScopeKind = 'SERVER' | 'GROUP' | 'ROOM';

export type PermissionScope =
  | { tier: 'server' }
  | { tier: 'group'; groupId: string }
  | { tier: 'room'; roomId: string };

export type TierPermissions = {
  permissions: string[];
  permissionDenials: string[];
};

export type TierRole = {
  roleName: string;
  displayName: string;
  description: string;
  isSystem: boolean;
  position: number;
  override: TierPermissions;
  inheritedAllows: string[];
  inheritedDenials: string[];
};

export type TierRoles = {
  applicablePermissions: string[];
  roles: TierRole[];
};

export type MatrixScope = {
  id: string;
  label: string;
  kind: MatrixScopeKind;
  parentGroupId: string;
};

export type MatrixCell = {
  permission: string;
  scopeId: string;
  override: MatrixDecision;
  effective: MatrixDecision;
};

export type MatrixData = {
  applicablePermissions: string[];
  scopes: MatrixScope[];
  cells: MatrixCell[];
};

export type RolePermissionMatrix = MatrixData & {
  roleName: string;
};

export type UserPermissionMatrix = MatrixData & {
  userId: string;
};

export function createPermissionAPI(config: PermissionAPIConfig) {
  const transport = createConnectTransport({
    baseUrl: config.baseUrl,
    useBinaryFormat: true
  });
  const client = createClient(AdminPermissionService, transport);
  const headers = () =>
    config.bearerToken ? { Authorization: `Bearer ${config.bearerToken}` } : undefined;

  return {
    async getRolePermissionTierMatrix(input: {
      roomId?: string | null;
      groupId?: string | null;
    }): Promise<TierRoles | null> {
      const response = await client.getRolePermissionTierMatrix(
        {
          scope: apiTierMatrixScope(input)
        },
        { headers: headers() }
      );
      return response.matrix ? tierRoles(response.matrix) : null;
    },

    async getRolePermissionMatrix(roleName: string): Promise<RolePermissionMatrix | null> {
      const response = await client.getRolePermissionMatrix({ roleName }, { headers: headers() });
      return response.matrix ? rolePermissionMatrix(response.matrix) : null;
    },

    async getUserPermissionMatrix(userId: string): Promise<UserPermissionMatrix | null> {
      const response = await client.getUserPermissionMatrix({ userId }, { headers: headers() });
      return response.matrix ? userPermissionMatrix(response.matrix) : null;
    },

    async setRolePermission(input: {
      roleName: string;
      scope: PermissionScope;
      permission: string;
      state: PermissionState;
    }): Promise<boolean> {
      const response = await client.setRolePermission(
        {
          roleName: input.roleName,
          permission: input.permission,
          decision: apiDecision(input.state),
          scope: apiScope(input.scope)
        },
        { headers: headers() }
      );
      return response.ok;
    },

    async setUserPermission(input: {
      userId: string;
      scope: PermissionScope;
      permission: string;
      state: PermissionState;
    }): Promise<boolean> {
      const response = await client.setUserPermission(
        {
          userId: input.userId,
          permission: input.permission,
          decision: apiDecision(input.state),
          scope: apiScope(input.scope)
        },
        { headers: headers() }
      );
      return response.ok;
    }
  };
}

export type PermissionAPI = ReturnType<typeof createPermissionAPI>;

function tierRoles(matrix: APITierRoles): TierRoles {
  return {
    applicablePermissions: [...matrix.applicablePermissions],
    roles: matrix.roles.map(tierRole)
  };
}

function tierRole(role: APITierRole): TierRole {
  return {
    roleName: role.roleName,
    displayName: role.displayName,
    description: role.description,
    isSystem: role.isSystem,
    position: role.position,
    override: {
      permissions: [...(role.override?.permissions ?? [])],
      permissionDenials: [...(role.override?.permissionDenials ?? [])]
    },
    inheritedAllows: [...role.inheritedAllows],
    inheritedDenials: [...role.inheritedDenials]
  };
}

function rolePermissionMatrix(matrix: APIRolePermissionMatrix): RolePermissionMatrix {
  return {
    roleName: matrix.roleName,
    applicablePermissions: [...matrix.applicablePermissions],
    scopes: matrix.scopes.map(matrixScope),
    cells: matrix.cells.map(matrixCell)
  };
}

function userPermissionMatrix(matrix: APIUserPermissionMatrix): UserPermissionMatrix {
  return {
    userId: matrix.userId,
    applicablePermissions: [...matrix.applicablePermissions],
    scopes: matrix.scopes.map(matrixScope),
    cells: matrix.cells.map(matrixCell)
  };
}

function matrixScope(scope: APIPermissionMatrixScope): MatrixScope {
  return {
    id: scope.id,
    label: scope.label,
    kind: scopeKind(scope.kind),
    parentGroupId: scope.parentGroupId
  };
}

function matrixCell(cell: APIPermissionMatrixCell): MatrixCell {
  return {
    permission: cell.permission,
    scopeId: cell.scopeId,
    override: matrixDecision(cell.override),
    effective: matrixDecision(cell.effective)
  };
}

function scopeKind(kind: PermissionScopeKind): MatrixScopeKind {
  if (kind === PermissionScopeKind.GROUP) return 'GROUP';
  if (kind === PermissionScopeKind.ROOM) return 'ROOM';
  return 'SERVER';
}

function matrixDecision(decision: PermissionDecision): MatrixDecision {
  if (decision === PermissionDecision.ALLOW) return 'ALLOW';
  if (decision === PermissionDecision.DENY) return 'DENY';
  return 'NONE';
}

function apiDecision(state: PermissionState): PermissionDecision {
  if (state === 'allow') return PermissionDecision.ALLOW;
  if (state === 'deny') return PermissionDecision.DENY;
  return PermissionDecision.NONE;
}

function apiScope(scope: PermissionScope): { kind: PermissionScopeKind; id: string } {
  if (scope.tier === 'group') {
    return { kind: PermissionScopeKind.GROUP, id: scope.groupId };
  }
  if (scope.tier === 'room') {
    return { kind: PermissionScopeKind.ROOM, id: scope.roomId };
  }
  return { kind: PermissionScopeKind.SERVER, id: '' };
}

function apiTierMatrixScope(input: {
  roomId?: string | null;
  groupId?: string | null;
}): { kind: PermissionScopeKind; id: string } {
  if (input.roomId) {
    return { kind: PermissionScopeKind.ROOM, id: input.roomId };
  }
  if (input.groupId) {
    return { kind: PermissionScopeKind.GROUP, id: input.groupId };
  }
  return { kind: PermissionScopeKind.SERVER, id: '' };
}

File: cli/internal/connectapi/permissions.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

type permissionService struct {
	api *API
}

func (s *permissionService) GetRolePermissionTierMatrix(ctx context.Context, req *connect.Request[apiv1.GetRolePermissionTierMatrixRequest]) (*connect.Response[apiv1.GetRolePermissionTierMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	roomID, groupID, err := permissionScopeIDs(req.Msg.GetScope())
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetRolePermissionTierMatrix(ctx, caller.UserID, roomID, groupID)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetRolePermissionTierMatrixResponse{Matrix: apiTierRoles(matrix)}), nil
}

func (s *permissionService) GetRolePermissionMatrix(ctx context.Context, req *connect.Request[apiv1.GetRolePermissionMatrixRequest]) (*connect.Response[apiv1.GetRolePermissionMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetRolePermissionMatrix(ctx, caller.UserID, req.Msg.GetRoleName())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetRolePermissionMatrixResponse{Matrix: apiRolePermissionMatrix(matrix)}), nil
}

func (s *permissionService) GetUserPermissionMatrix(ctx context.Context, req *connect.Request[apiv1.GetUserPermissionMatrixRequest]) (*connect.Response[apiv1.GetUserPermissionMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetUserPermissionMatrix(ctx, caller.UserID, req.Msg.GetUserId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.GetUserPermissionMatrixResponse{Matrix: apiUserPermissionMatrix(matrix)}), nil
}

func (s *permissionService) ExplainPermissions(ctx context.Context, req *connect.Request[apiv1.ExplainPermissionsRequest]) (*connect.Response[apiv1.ExplainPermissionsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	explanations, err := s.api.core.ExplainPermissions(ctx, caller.UserID, req.Msg.GetUserId(), req.Msg.GetRoomId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.ExplainPermissionsResponse{Explanations: apiPermissionExplanations(explanations)}), nil
}

func (s *permissionService) SetRolePermission(ctx context.Context, req *connect.Request[apiv1.SetRolePermissionRequest]) (*connect.Response[apiv1.SetRolePermissionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	state, err := corePermissionState(req.Msg.GetDecision())
	if err != nil {
		return nil, err
	}
	scope, err := corePermissionTargetScope(req.Msg.GetScope())
	if err != nil {
		return nil, err
	}
	if err := s.api.core.SetRolePermissionState(ctx, caller.UserID, req.Msg.GetRoleName(), scope, core.Permission(req.Msg.GetPermission()), state); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.SetRolePermissionResponse{Ok: true}), nil
}

func (s *permissionService) SetUserPermission(ctx context.Context, req *connect.Request[apiv1.SetUserPermissionRequest]) (*connect.Response[apiv1.SetUserPermissionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	state, err := corePermissionState(req.Msg.GetDecision())
	if err != nil {
		return nil, err
	}
	scope, err := corePermissionTargetScope(req.Msg.GetScope())
	if err != nil {
		return nil, err
	}
	if err := s.api.core.SetUserPermissionState(ctx, caller.UserID, req.Msg.GetUserId(), scope, core.Permission(req.Msg.GetPermission()), state); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.SetUserPermissionResponse{Ok: true}), nil
}

func apiPermissionExplanations(explanations []core.PermissionExplanation) []*apiv1.PermissionExplanation {
	out := make([]*apiv1.PermissionExplanation, 0, len(explanations))
	for i := range explanations {
		out = append(out, apiPermissionExplanation(explanations[i]))
	}
	return out
}

func apiPermissionExplanation(explanation core.PermissionExplanation) *apiv1.PermissionExplanation {
	out := &apiv1.PermissionExplanation{
		Permission:    string(explanation.Permission),
		State:         apiPermissionExplanationDecision(explanation.State),
		DecidedAt:     apiPermissionDecisionLevel(explanation.DecidedAt),
		DecidedByRole: explanation.DecidedByRole,
		Trace:         make([]*apiv1.PermissionTraceEntry, 0, len(explanation.Trace)),
	}
	for i, entry := range explanation.Trace {
		out.Trace = append(out.Trace, &apiv1.PermissionTraceEntry{
			Level:    apiPermissionDecisionLevel(entry.Level),
			RoleName: entry.RoleName,
			Decision: apiPermissionExplanationDecision(entry.Decision),
			Applied:  i == 0,
		})
	}
	return out
}

func apiPermissionExplanationDecision(decision core.DecisionKind) apiv1.PermissionDecision {
	switch decision {
	case core.DecisionAllow:
		return apiv1.PermissionDecision_PERMISSION_DECISION_ALLOW
	case core.DecisionDeny:
		return apiv1.PermissionDecision_PERMISSION_DECISION_DENY
	default:
		return apiv1.PermissionDecision_PERMISSION_DECISION_NONE
	}
}

func apiPermissionDecisionLevel(level core.PermissionLevel) apiv1.PermissionDecisionLevel {
	switch level {
	case core.LevelServer:
		return apiv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_SERVER
	case core.LevelGroup:
		return apiv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_GROUP
	case core.LevelRoom:
		return apiv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_ROOM
	default:
		return apiv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_UNSPECIFIED
	}
}

func apiTierRoles(matrix *core.TierRoles) *apiv1.TierRoles {
	if matrix == nil {
		return nil
	}
	out := &apiv1.TierRoles{
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Roles:                 make([]*apiv1.TierRole, 0, len(matrix.Roles)),
	}
	for _, role := range matrix.Roles {
		out.Roles = append(out.Roles, &apiv1.TierRole{
			RoleName:         role.RoleName,
			DisplayName:      role.DisplayName,
			Description:      role.Description,
			IsSystem:         role.IsSystem,
			Position:         role.Position,
			Override:         apiTierPermissions(role.Override),
			InheritedAllows:  append([]string(nil), role.InheritedAllows...),
			InheritedDenials: append([]string(nil), role.InheritedDenials...),
		})
	}
	return out
}

func apiTierPermissions(perms core.TierPermissions) *apiv1.TierPermissions {
	return &apiv1.TierPermissions{
		Permissions:       append([]string(nil), perms.Permissions...),
		PermissionDenials: append([]string(nil), perms.PermissionDenials...),
	}
}

func apiRolePermissionMatrix(matrix *core.RolePermissionMatrix) *apiv1.RolePermissionMatrix {
	if matrix == nil {
		return nil
	}
	return &apiv1.RolePermissionMatrix{
		RoleName:              matrix.RoleName,
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Scopes:                apiPermissionMatrixScopes(matrix.Scopes),
		Cells:                 apiPermissionMatrixCells(matrix.Cells),
	}
}

func apiUserPermissionMatrix(matrix *core.UserPermissionMatrix) *apiv1.UserPermissionMatrix {
	if matrix == nil {
		return nil
	}
	return &apiv1.UserPermissionMatrix{
		UserId:                matrix.UserID,
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Scopes:                apiPermissionMatrixScopes(matrix.Scopes),
		Cells:                 apiPermissionMatrixCells(matrix.Cells),
	}
}

func apiPermissionMatrixScopes(scopes []core.PermissionMatrixScope) []*apiv1.PermissionMatrixScope {
	out := make([]*apiv1.PermissionMatrixScope, 0, len(scopes))
	for _, scope := range scopes {
		out = append(out, &apiv1.PermissionMatrixScope{
			Id:            scope.ID,
			Label:         scope.Label,
			Kind:          apiPermissionScopeKind(scope.Kind),
			ParentGroupId: scope.ParentGroupID,
		})
	}
	return out
}

func apiPermissionMatrixCells(cells []core.PermissionMatrixCell) []*apiv1.PermissionMatrixCell {
	out := make([]*apiv1.PermissionMatrixCell, 0, len(cells))
	for _, cell := range cells {
		out = append(out, &apiv1.PermissionMatrixCell{
			Permission: cell.Permission,
			ScopeId:    cell.ScopeID,
			Override:   apiPermissionDecision(cell.Override),
			Effective:  apiPermissionDecision(cell.Effective),
		})
	}
	return out
}

func apiPermissionScopeKind(kind core.MatrixScopeKind) apiv1.PermissionScopeKind {
	switch kind {
	case core.MatrixScopeGroup:
		return apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_GROUP
	case core.MatrixScopeRoom:
		return apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_ROOM
	default:
		return apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_SERVER
	}
}

func apiPermissionDecision(decision core.MatrixDecision) apiv1.PermissionDecision {
	switch decision {
	case core.MatrixDecisionAllow:
		return apiv1.PermissionDecision_PERMISSION_DECISION_ALLOW
	case core.MatrixDecisionDeny:
		return apiv1.PermissionDecision_PERMISSION_DECISION_DENY
	default:
		return apiv1.PermissionDecision_PERMISSION_DECISION_NONE
	}
}

func corePermissionState(decision apiv1.PermissionDecision) (core.PermissionState, error) {
	switch decision {
	case apiv1.PermissionDecision_PERMISSION_DECISION_ALLOW:
		return core.PermissionStateAllow, nil
	case apiv1.PermissionDecision_PERMISSION_DECISION_DENY:
		return core.PermissionStateDeny, nil
	case apiv1.PermissionDecision_PERMISSION_DECISION_NONE:
		return core.PermissionStateNone, nil
	default:
		return "", invalidArgument("decision is required")
	}
}

func permissionScopeIDs(scope *apiv1.PermissionScope) (roomID string, groupID string, err error) {
	if scope == nil {
		return "", "", nil
	}
	switch scope.GetKind() {
	case apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_UNSPECIFIED:
		if scope.GetId() != "" {
			return "", "", invalidArgument("server scope id must be empty")
		}
		return "", "", nil
	case apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_SERVER:
		if scope.GetId() != "" {
			return "", "", invalidArgument("server scope id must be empty")
		}
		return "", "", nil
	case apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_GROUP:
		if scope.GetId() == "" {
			return "", "", invalidArgument("group scope id is required")
		}
		return "", scope.GetId(), nil
	case apiv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_ROOM:
		if scope.GetId() == "" {
			return "", "", invalidArgument("room scope id is required")
		}
		return scope.GetId(), "", nil
	default:
		return "", "", invalidArgument("unsupported permission scope kind")
	}
}

func corePermissionTargetScope(scope *apiv1.PermissionScope) (core.PermissionTargetScope, error) {
	roomID, groupID, err := permissionScopeIDs(scope)
	if err != nil {
		return core.PermissionTargetScope{}, err
	}
	switch {
	case roomID != "":
		return core.PermissionTargetScope{Kind: core.MatrixScopeRoom, ID: roomID}, nil
	case groupID != "":
		return core.PermissionTargetScope{Kind: core.MatrixScopeGroup, ID: groupID}, nil
	default:
		return core.PermissionTargetScope{}, nil
	}
}
